pkg/destroy/powervs/serviceinstance.go

```diff
@@ -0,0 +1,281 @@
+		}
+
+		moreData = *resources.RowsCount == perPage
+	}
+
+	return cloudResources{}.insert(result...), nil
+}
+
+// destroyServiceInstance destroys a service instance.
+func (o *ClusterUninstaller) destroyServiceInstance(item cloudResource) error {
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	select {
+	case <-ctx.Done():
+		o.Logger.Debugf("destroyServiceInstance: case <-ctx.Done()")
+		return o.Context.Err() // we're cancelled, abort
+	default:
+	}
+
+	o.Logger.Debugf("destroyServiceInstance: Preparing to delete, item.name = %v", item.name)
+
+	var (
+		getOptions *resourcecontrollerv2.GetResourceInstanceOptions
+		response   *core.DetailedResponse
+		err        error
+	)
+
+	getOptions = o.controllerSvc.NewGetResourceInstanceOptions(item.id)
+
+	_, response, err = o.controllerSvc.GetResourceInstanceWithContext(ctx, getOptions)
+
+	if err != nil && response != nil && response.StatusCode == gohttp.StatusNotFound {
+		// The resource is gone
+		o.deletePendingItems(item.typeName, []cloudResource{item})
+		o.Logger.Infof("Deleted Service Instance %q", item.name)
+		return nil
+	}
+	if err != nil && response != nil && response.StatusCode == gohttp.StatusInternalServerError {
+		o.Logger.Infof("destroyServiceInstance: internal server error")
+		return nil
+	}
+
+	options := o.controllerSvc.NewDeleteResourceInstanceOptions(item.id)
+	options.SetRecursive(true)
+
+	response, err = o.controllerSvc.DeleteResourceInstanceWithContext(ctx, options)
+
+	if err != nil && response != nil && response.StatusCode != gohttp.StatusNotFound {
+		return errors.Wrapf(err, "failed to delete service instance %s", item.name)
+	}
+
+	o.Logger.Infof("Deleted Service Instance %q", item.name)
+	o.deletePendingItems(item.typeName, []cloudResource{item})
+
+	return nil
+}
+
+// destroyServiceInstances removes all service instances have a name containing
+// the cluster's infra ID.
+func (o *ClusterUninstaller) destroyServiceInstances() error {
+	firstPassList, err := o.listServiceInstances()
+	if err != nil {
+		return err
+	}
+
+	if len(firstPassList.list()) == 0 {
+		return nil
+	}
+
+	items := o.insertPendingItems(serviceInstanceTypeName, firstPassList.list())
+
+	ctx, cancel := o.contextWithTimeout()
+	defer cancel()
+
+	for _, item := range items {
+		select {
+		case <-ctx.Done():
+			o.Logger.Debugf("destroyServiceInstances: case <-ctx.Done()")
+			return o.Context.Err() // we're cancelled, abort
+		default:
+		}
+
+		backoff := wait.Backoff{
+			Duration: 15 * time.Second,
+			Factor:   1.1,
+			Cap:      leftInContext(ctx),
+			Steps:    math.MaxInt32}
+		err = wait.ExponentialBackoffWithContext(ctx, backoff, func(context.Context) (bool, error) {
+			err2 := o.destroyServiceInstance(item)
+			if err2 == nil {
+				return true, err2
+			}
+			o.errorTracker.suppressWarning(item.key, err2, o.Logger)
+			return false, err2
+		})
+		if err != nil {
+			o.Logger.Fatal("destroyServiceInstances: ExponentialBackoffWithContext (destroy) returns ", err)
```

Comment:
`Fatal` calls `os.Exit(1)` after logging ([source](https://github.com/sirupsen/logrus#level-logging)) is that what you intend here?

Reply:
Yes.  The destroy code everywhere expects to work perfectly or fail immediately and not continue destroying things.